nodeInterface/virtual-contracts.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package nodeInterface
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/arbitrum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbnode"
+	"github.com/offchainlabs/nitro/arbos/arbosState"
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/precompiles"
+	"github.com/offchainlabs/nitro/solgen/go/node_interfacegen"
+	"github.com/offchainlabs/nitro/solgen/go/precompilesgen"
+	"github.com/offchainlabs/nitro/util/arbmath"
+)
+
+type addr = common.Address
+type mech = *vm.EVM
+type huge = *big.Int
+type hash = common.Hash
+type bytes32 = [32]byte
+type ctx = *precompiles.Context
+
+type Message = types.Message
+type BackendAPI = core.NodeInterfaceBackendAPI
+type ExecutionResult = core.ExecutionResult
+
+func init() {
+	arbstate.RequireHookedGeth()
+
+	nodeInterfaceImpl := &NodeInterface{Address: types.NodeInterfaceAddress}
+	nodeInterfaceMeta := node_interfacegen.NodeInterfaceMetaData
+	_, nodeInterface := precompiles.MakePrecompile(nodeInterfaceMeta, nodeInterfaceImpl)
+
+	nodeInterfaceDebugImpl := &NodeInterfaceDebug{Address: types.NodeInterfaceDebugAddress}
+	nodeInterfaceDebugMeta := node_interfacegen.NodeInterfaceDebugMetaData
+	_, nodeInterfaceDebug := precompiles.MakePrecompile(nodeInterfaceDebugMeta, nodeInterfaceDebugImpl)
+
+	resetInterfaceImpls := func(msg Message, ret *Message, retBool *bool, ctx context.Context, backend BackendAPI) {
+		nodeInterfaceImpl.backend = backend
```

Comment:
I think this presents concurrency issues. We need to have a different NodeInterface struct for each RPC invocation.

Reply:
Fixed. Each call creates a unique copy of the impl to avoid concurrency issues